File: examples/excel-number-format.ts
/**
 * 例子（Epic-44）：演示 setBuiltInNumberFormat（货币 / 百分比 / 日期 内建格式）。
 *
 * 跑法：
 *   bun run examples/excel-number-format.ts <output.xlsx>
 */

import { StringValue, UInt32Value } from "../src/element/index.js";
import {
  BuiltInNumberFormat,
  Cell,
  CellValue,
  Row,
  SheetData,
  SpreadsheetDocument,
  setBuiltInNumberFormat,
} from "../src/excel/index.js";

async function main(): Promise<void> {
  const [outputPath] = process.argv.slice(2);
  if (outputPath === undefined) {
    process.stderr.write("usage: excel-number-format <output.xlsx>\n");
    process.exit(2);
  }

  const doc = SpreadsheetDocument.create();
  const ws = doc.workbookPart!.worksheetParts[0]!.worksheet;
  const sheetData = ws.firstChild(SheetData);
  if (sheetData === undefined) throw new Error("missing SheetData");

  function addRow(idx: number, fmt: number | undefined, raw: string, label: string): void {
    const row = new Row();
    row.rowIndex = UInt32Value.parse(String(idx));
    const labelCell = new Cell();
    labelCell.cellReference = StringValue.parse(`A${idx}`);
    labelCell.dataType = StringValue.parse("str");
    const lv = new CellValue();
    lv.text = label;
    labelCell.appendChild(lv);
    row.appendChild(labelCell);

    const valueCell = new Cell();
    valueCell.cellReference = StringValue.parse(`B${idx}`);
    const v = new CellValue();
    v.text = raw;
    valueCell.appendChild(v);
    if (fmt !== undefined) setBuiltInNumberFormat(doc, valueCell, fmt);
    row.appendChild(valueCell);
    sheetData.appendChild(row);
  }

  addRow(1, BuiltInNumberFormat.INTEGER, "1234", "Integer:");
  addRow(2, BuiltInNumberFormat.DECIMAL_2, "1234.5", "2 decimals:");
  addRow(3, BuiltInNumberFormat.THOUSANDS_DECIMAL_2, "1234567.89", "Thousands+2:");
  addRow(4, BuiltInNumberFormat.PERCENT_DECIMAL_2, "0.1234", "Percent:");
  addRow(5, BuiltInNumberFormat.CURRENCY, "1234.5", "Currency:");
  addRow(6, BuiltInNumberFormat.DATE_SHORT, "45431", "Date (serial):");

  await doc.saveAsAsync(outputPath);
  process.stdout.write(`Wrote ${outputPath} (6 rows of built-in number formats)\n`);
}

main().catch((err) => {
  process.stderr.write(`failed: ${(err as Error).stack ?? err}\n`);
  process.exit(1);
});

File: tools/cross-sdk-verify/replicas/ExcelNumberFormat.cs
// Replica of examples/excel-number-format.ts
// Creates xlsx with 6 rows demonstrating built-in number formats.

using DocumentFormat.OpenXml;
using DocumentFormat.OpenXml.Packaging;
using DocumentFormat.OpenXml.Spreadsheet;

namespace CrossSdkVerify.Replicas;

public static class ExcelNumberFormat
{
    // Built-in number format IDs matching BuiltInNumberFormat in the TS lib:
    // INTEGER=1, DECIMAL_2=2, THOUSANDS_DECIMAL_2=4, PERCENT_DECIMAL_2=10, CURRENCY=44, DATE_SHORT=14
    private const uint FmtInteger = 1;
    private const uint FmtDecimal2 = 2;
    private const uint FmtThousandsDecimal2 = 4;
    private const uint FmtPercentDecimal2 = 10;
    private const uint FmtCurrency = 44;
    private const uint FmtDateShort = 14;

    public static void Run(string outputPath)
    {
        using var doc = SpreadsheetDocument.Create(outputPath,
            DocumentFormat.OpenXml.SpreadsheetDocumentType.Workbook);

        var wbPart = doc.AddWorkbookPart();
        wbPart.Workbook = new Workbook();

        // Stylesheet with cell formats referencing built-in number formats
        var stylesheet = BuildStylesheet();
        var stylePart = wbPart.AddNewPart<WorkbookStylesPart>();
        stylePart.Stylesheet = stylesheet;
        stylePart.Stylesheet.Save();

        var wsPart = wbPart.AddNewPart<WorksheetPart>();
        var sheetData = new SheetData();
        wsPart.Worksheet = new Worksheet(sheetData);

        var sheets = wbPart.Workbook.AppendChild(new Sheets());
        sheets.Append(new Sheet
        {
            Id = wbPart.GetIdOfPart(wsPart),
            SheetId = 1,
            Name = "Sheet1"
        });

        // Row data: (label, raw value, style index)
        // Style indices match the order in BuildStylesheet CellFormats:
        // 0=default, 1=INTEGER, 2=DECIMAL_2, 3=THOUSANDS_DECIMAL_2,
        // 4=PERCENT_DECIMAL_2, 5=CURRENCY, 6=DATE_SHORT
        AddRow(sheetData, 1, "Integer:", "1234", 1);
        AddRow(sheetData, 2, "2 decimals:", "1234.5", 2);
        AddRow(sheetData, 3, "Thousands+2:", "1234567.89", 3);
        AddRow(sheetData, 4, "Percent:", "0.1234", 4);
        AddRow(sheetData, 5, "Currency:", "1234.5", 5);
        AddRow(sheetData, 6, "Date (serial):", "45431", 6);

        wbPart.Workbook.Save();
    }

    private static void AddRow(SheetData sd, int rowIdx, string label, string raw, uint styleIdx)
    {
        var row = new Row { RowIndex = (uint)rowIdx };

        // A: label (inline string)
        row.Append(new Cell
        {
            CellReference = $"A{rowIdx}",
            DataType = CellValues.String,
            CellValue = new CellValue(label)
        });

        // B: numeric value with number format style
        row.Append(new Cell
        {
            CellReference = $"B{rowIdx}",
            CellValue = new CellValue(raw),
            StyleIndex = styleIdx
        });

        sd.Append(row);
    }

    private static Stylesheet BuildStylesheet()
    {
        // We need one CellFormat per row (6 formats + 1 default = 7 total)
        var cellFormats = new CellFormats();
        cellFormats.Append(new CellFormat()); // 0: default
        cellFormats.Append(new CellFormat { NumberFormatId = FmtInteger, ApplyNumberFormat = true });       // 1
        cellFormats.Append(new CellFormat { NumberFormatId = FmtDecimal2, ApplyNumberFormat = true });      // 2
        cellFormats.Append(new CellFormat { NumberFormatId = FmtThousandsDecimal2, ApplyNumberFormat = true }); // 3
        cellFormats.Append(new CellFormat { NumberFormatId = FmtPercentDecimal2, ApplyNumberFormat = true }); // 4
        cellFormats.Append(new CellFormat { NumberFormatId = FmtCurrency, ApplyNumberFormat = true });     // 5
        cellFormats.Append(new CellFormat { NumberFormatId = FmtDateShort, ApplyNumberFormat = true });    // 6

        return new Stylesheet(
            new Fonts(new Font()),
            new Fills(new Fill(new PatternFill { PatternType = PatternValues.None })),
            new Borders(new Border()),
            cellFormats
        );
    }
}
